### ISSUE TITLE
After AzureAD outages MSAL cookie becomes corrupt

### ISSUE THREAD

**Issue by imakeinterfaces** [AzureAD/microsoft-authentication-library-for-js#3533]
### Core Library

@azure/msal-browser

### Core Library Version

2.11.2

### Wrapper Library

Not Applicable

### Wrapper Library Version

2.11.2

### Description

We have noticed in the past that during Microsoft AzureAD outages, the cookie state of our users becomes corrupt necessitating a cookie reset. This is a widespread problem that affects all users who were using our application during the outage.

Our setup uses the msal-browser JS library directly, not the react or angular specific versions.

When this happens, we log 2 different errors in our system.

1. invalid_grant: AADSTS50078: Presented multi-factor authentication has expired due to policies configured by your administrator, you must refresh your multi-factor authentication to access 'xxx-xxx-xxx-xxx-xxx'. Trace ID: xxx-cd68-xxx-xxx-xxx Correlation ID: xxx-xxx-xxx-xxx-xxx Timestamp: 2021-04-25 12:14:13Z

2. Unhandled promise rejection: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.

Our users end up being locked out because their `interaction_in_progress` cookie key becomes stuck in the true state even though the redirect failed. I have gone so far as to create a logout button that manually removes this cookie key when we catch an error. 

> 

### Error Message

invalid_grant: AADSTS50078: Presented multi-factor authentication has expired due to policies configured by your administrator, you must refresh your multi-factor authentication to access 'xxx-xxx-xxx-xxx-xxx'. Trace ID: 6bc6d06a-xxx-xxx-xxx-xxx Correlation ID: xxx-xxx-xxx-xxx-xxx Timestamp: 2021-04-25 12:14:13Z
src.cbec75e9.js in t at line 1594
src.cbec75e9.js in t at line 1594
src.cbec75e9.js in t at line 1594
src.cbec75e9.js in validateTokenResponse at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in a at line 1594

### Msal Logs

Unhandled promise rejection: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.
src.cbec75e9.js in t [as constructor] at line 1594
src.cbec75e9.js in new r at line 1594
src.cbec75e9.js in createInteractionInProgressError at line 1594
src.cbec75e9.js in preflightInteractiveRequest at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in next at line 1594
src.cbec75e9.js in <anonymous> at line 1594
src.cbec75e9.js in n at line 1594

### MSAL Configuration

```javascript
auth: {
    clientId: process.env.AZURE_CLIENT_ID,
    redirectUri: process.env.AZURE_REDIRECT_URI,
    navigateToLoginRequestUrl: false,
    authority: `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}`,
  },
  cache: {
    cacheLocation: 'localStorage',
    storeAuthStateInCookie: true,
  },
```

### Relevant Code Snippets

```javascript
Code snapshot:

function acquireToken (requestObject) {
  const ONE_DAY_SECONDS = 60 * 60 * 24;
  const secureCookie = process.env.SECURE_COOKIE_VALUE ? 'secure;' : '';

  global.PublicClientApplication.acquireTokenSilent(requestObject)
    .then(accessTokenResponse => {
      global.document.cookie = [
        `__myappauth=${accessTokenResponse.accessToken};`,
        `domain=${process.env.DOMAIN_NAME};`,
        `max-age=${ONE_DAY_SECONDS};`,
        `SameSite=lax;${secureCookie};`,
        `path=/`
      ].join('');
    })
    .catch(error => {
      global.PublicClientApplication.loginRedirect();
      global.apm.captureError(error);
    });
}

function getUser(registerUser) {
  const currentAccounts = global.PublicClientApplication.getAllAccounts();

  if (currentAccounts && currentAccounts.length) {
    const userName = currentAccounts[0].username;
    registerUser({ variables: { email: userName } });
    global.PublicClientApplication.setActiveAccount(currentAccounts[0]);
    return currentAccounts[0];
  }
  return null;
}

function requireAuth (history, pagePermissions) {
    const cachedUser = getUser(registerUser);
    if (cachedUser) {
      acquireToken({
        scopes: [process.env.AZURE_API_SCOPE],
        account: cachedUser,
      });
    } else {
      global.PublicClientApplication.handleRedirectPromise()
        .then(resp => {
          if (resp !== null) {
            global.PublicClientApplication.setActiveAccount(resp.account);
          } else {
            getUser();
          }
        })
        .then(result => {
          const newUser = getUser(registerUser);
          if (!result && !newUser) {
            history.replace({
              pathname: '/login',
              state: { nextPathname: history.pathname },
            });
          } else {
            acquireToken({
              scopes: [process.env.AZURE_API_SCOPE],
              account: newUser,
            }); 
          }
        })
        .catch(error => {
          externalLogger.log(err)
          global.PublicClientApplication.loginRedirect();
        });
    }
  }

// force logout method mentioned above
  function logout() {
    // If two microsoft accounts are logged in, sometimes the interaction status cookie will get stuck true
    Cookies.remove(`msal.${process.env.AZURE_CLIENT_ID}.interaction.status`);
    global.PublicClientApplication.logout();
  }
```

### Reproduction Steps

1. Use azureAD during a microsoft outage
2. The affected user will no longer be able to log in, even after service has been restored.
3. Clear the user's cookie with key `msal.xxx.interaction_in_progress` will allow the user to log back in again. Though we recommend most users reset their browser from the chrome settings since they are not developers.

### Expected Behavior

Client state should not become corrupted during outages.

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome, Firefox, Safari

### Source

External (Customer)

**Comment by jasonnutter** [comment]
@imakeinterfaces Please upgrade to the latest version of MSAL Browse (2.14.1) and see if this is still an issue, as we made fixes in 2.12.1 to properly clear temporary cache values: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/changelog.md#2121

**Comment by ascott18** [comment]
@jasonnutter I am on 2.14.1 and am having the same issue. If I put a breakpoint in the clearMsalCookies, it is never hit.

**Comment by jasonnutter** [comment]
> @jasonnutter I am on 2.14.1 and am having the same issue. If I put a breakpoint in the clearMsalCookies, it is never hit.

@ascott18 After invoking which API, logout or acquireTokenSilent? Or something else?

**Comment by ascott18** [comment]
@jasonnutter This is on a page that is using `loginRedirect`. If the user closes the tab without completing the authentication flow, our app is bricked for that user and is no longer able to perform `loginRedirect` anymore.

**Comment by jasonnutter** [comment]
> @jasonnutter This is on a page that is using `loginRedirect`. If the user closes the tab without completing the authentication flow, our app is bricked for that user and is no longer able to perform `loginRedirect` anymore.

@ascott18 I see, this should be handled by `handleRedirectPromise`: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/26c54cc4e186ae1164a73a5155f9e853ba173153/lib/msal-browser/src/app/ClientApplication.ts#L202-L208

Can you debug `handleRedirectPromise` and see what is happening there? I'll also see if I can repro.

**Comment by ascott18** [comment]
@jasonnutter Thanks, adding a call to handleRedirectPromise at the beginning of our auth flow resolved this. This is very unintuitive and makes it very difficult to fall into the "pit of success" when using this library. Why is it designed such that we have to call this function even on pages that are not the target of a redirect? At best, this is a dire violation of the Single Responsibility Principal.

**Comment by jasonnutter** [comment]
> @jasonnutter Thanks, adding a call to handleRedirectPromise at the beginning of our auth flow resolved this. This is very unintuitive and makes it very difficult to fall into the "pit of success" when using this library. Why is it designed such that we have to call this function even on pages that are not the target of a redirect? At best, this is a dire violation of the Single Responsibility Principal.

In MSAL.js v1, the constructor of `UserAgentApplication` automatically handled both redirect logic processing and cleanup, however, we received feedback that this was too opinionated (and also violated the single responsibility principal). We decided to move this logic to `handleRedirectPromise` and required applications to invoke it on pages used for redirect URIs (and any other pages that may initiate a redirect). This was done to address that feedback and to account for the fact that in MSAL.js v2, redirect processing is asynchronous, so we could not include redirect processing in the constructor. 

Furthermore, `handleRedirectPromise` must be invoked on pages that initiate redirects because with `navigateToLoginRequestURl` set to true, we decided that we wanted applications to receive the redirect response on the page that initiated the redirect (as it was previously unintuitive when the redirect uri would get the redirect response, but then immediately get redirect to the redirect start page, which would not get the response returned to it).

For the next version of MSAL.js, we are strongly considering adding a "startup" function that all applications must invoke (on all pages) \, which would potentially perform cleanup tasks such as the one we are talking about. We provide APIs like this as part of our goal to abstract away as much of the underlying implementation details and nuances of integrating with AAD, which still providing a composable set of APIs that can be used for as many scenarios and use cases as possible.

Let me know if you have further questions or feedback!
